xwiki-commons-core/xwiki-commons-job/src/main/java/org/xwiki/job/internal/JobGroupPathLockTree.java

```diff
@@ -32,31 +38,104 @@
  * @version $Id$
  * @since 6.1M2
  */
+@Component(roles = JobGroupPathLockTree.class)
+@Singleton
 public class JobGroupPathLockTree
 {
-    private final Map<JobGroupPath, ReadWriteLock> tree = new ConcurrentHashMap<>();
+    @Inject
+    private GroupedJobInitializerManager groupedJobInitializerManager;
+
+    /**
+     * A specific concurrency implementation for managing Semaphore with Read/Write lock capabilities.
+     */
+    private class ReadWriteSemaphore
+    {
+        private volatile int readCounter;
+        private volatile int writeCounter;
+        private Semaphore semaphore;
+        private Lock writeCounterLock;
+
+        ReadWriteSemaphore(int poolSize)
+        {
+            this.semaphore = new Semaphore(poolSize, true);
+            this.writeCounterLock = new ReentrantLock(true);
+            this.readCounter = 0;
+            this.writeCounter = 0;
+        }
+
+        void lockWrite()
+        {
+            this.writeCounterLock.lock();
+            this.writeCounter++;
+            this.writeCounterLock.unlock();
+            this.semaphore.acquireUninterruptibly(this.readCounter + 1);
+        }
+
+        void unlockWrite()
+        {
+            this.writeCounterLock.lock();
+            this.writeCounter--;
+
+            if (this.writeCounter == 0) {
+                this.writeCounterLock.unlock();
+                this.semaphore.release(this.readCounter + 1);
+            } else {
+                this.writeCounterLock.unlock();
+                this.semaphore.release();
+            }
+        }
+
+        void lockRead()
+        {
+            this.readCounter++;
```

Comment:
@tmortagne do you think I should protect them too with a dedicated lock, like I did for `writeCounter`? I started to do it, but then removed it after I got some trouble (deadlock) with them. I didn't put them back because I thought they were useless since we don't use that value for testing a condition, but we definitely use it for taking the semaphore permits, so now I'm hesitating about it...

Reply:
I guess I'd need to protect it, volatile won't protect them for "read-update-write" and so I could have a race condition with a lockRead/unlockRead leading to a wrong counter value.

Reply:
FYI I changed it to protect it.